### PR TITLE
Storage volumes all projects

### LINKED ISSUE #10753: Implement all-projects query parameter for storage volumes
In order to complete https://github.com/canonical/lxd-cloud/issues/136, upstream servers need to synchronise instance and storage volume records when the make a connection to a downstream. 

To accomplish this, we can call `/1.0/instances?all-projects=true` for instances, but there is currently no equivalent for volumes. We could iterate over projects and make a separate call for each but this would result in lots of unnecessary requests.

### PR DESCRIPTION
**Pull request by markylaing**
Implements an `all-projects=true` query parameter on `/1.0/storage-pools/{poolName}/volumes`. Adds associated methods to client and adds `--all-projects` flag to `lxc storage volume list {poolName}`.

Closes #10753

### CODE REVIEW

File: lxd/storage_volumes.go [hunk 1]
```diff
@@ -203,44 +205,53 @@ func storagePoolVolumesGet(d *Daemon, r *http.Request) response.Response {
...
-	if err != nil && !response.IsNotFoundError(err) {
-		return response.SmartError(err)
-	}
+	var volumes []*api.StorageVolume
+	if allProjects {
+		// Get all instance volumes currently attached to the storage pool by ID of the pool and project.
+		volumes, err = d.db.Cluster.GetStoragePoolVolumesAllProjects(poolID, supportedVolumeTypes)
+		if err != nil && !response.IsNotFoundError(err) {
+			return response.SmartError(err)
+		}
```
Comment (markylaing) [inline lxd/storage_volumes.go:214]: @tomponline I noticed there is some special handling of image volumes when getting volumes by project. `GetStorageVolumesAllProjects` will set the project of all image volumes to `default`. Do you think that is an issue?

**Review by tomponline — CHANGES_REQUESTED** [review]
Hi @markylaing I'm not super happy about this pull request, but not because of what you've added, but because of what is in the code currently.

In my view this particular API route is in dire need of clean up to make its DB query pattern more efficient (in order to be able to list hundreds of volumes without hitting database timeouts).

And your PR is building on this technical debt rather than alleviating it.
Your question about images is interesting, and needs exploration, I suspect returning them all as default will not suffice.

The way I would like us to solve this problem is first by making the existing API  route more efficient and clean, by implementing a `InstanceList()` type function that provides the ability to filter storage volumes by things like type, project, name (similar to GetInstances()) and then efficiently load the config and project info for the returned volumes in as few queries as possible.

By doing this properly we would then make adding the `all-projects` feature trivial as all of the underlying functionality would be there.

As this issue is blocking you on the other project, are you able to open an issue about this and then work around it in the other project by making 1 API call per project?

Thans

**Comment by tomponline** [comment]
I see you have already opened an issue, thanks.

**Comment by markylaing** [comment]
> In my view this particular API route is in dire need of clean up to make its DB query pattern more efficient (in order to be able to list hundreds of volumes without hitting database timeouts).

I noticed when working on this that there are no generated statements or methods for storage volumes but instead lots of methods defined on `db.Cluster` which initiate their own transaction per call. At this point I realised the refactoring would be quite extensive and simply listed projects.. which of course is not ideal.

Is there a reason why we didn't use `lxd-generate` for storage pools and volumes? If so, perhaps we should rewrite the storage pool and volume methods to follow a similar pattern.

> As this issue is blocking you on the other project, are you able to open an issue about this and then work around it in the other project by making 1 API call per project?

Yeah that's fine I'll make a call per project.

**Comment by tomponline** [comment]
> Is there a reason why we didn't use `lxd-generate` for storage pools and volumes? If so, perhaps we should rewrite the storage pool and volume methods to follow a similar pattern.

I'm not sure, although that is only part of the solution needed. Whilst we do need something that will allow us to load the storage volumes (without their associated configs and projects) in a filterable fashion (which would be where the DB generated functions would be useful). We do then also need something to interrogate the result set and then load the required configs and projects (and their configs) in specialised queries, and then provide that to the callback function so it can generate an api.StorageVolume struct if needed.

> Yeah that's fine I'll make a call per project.
Thanks

**Comment by tomponline** [comment]
@markylaing I'll take this one over as part of a storage volumes list refactor to add UsedBy support for storage buckets, if ok?

**Pull request closed without merging**